File: game_cpp/src/particle.cpp
#include "particle.h"



void particle::move(float dt)
{
	Vector dx = vel.MultiplyVectorByScalar(dt);
	part_loc = part_loc.AddVectorToPoint(&dx);
}

void particle::force_stop()
{
	vel = Vector(0, 0, 0);
}

void particle::accelerate(float dt)
{
	Vector dacc = acceleration.MultiplyVectorByScalar(dt);
	vel = vel.AddVectorToVector(&dacc);
	acceleration = Vector(0, 0, 0);

	for (int i = 0; i < attachements.size(); i++)
	{
		Vector dx = attachements[i].pt.SubtractPointFromPoint(&part_loc); //points from particle to att. pt.
		float distance = dx.get_size();
		if (distance >= attachements[i].length)
		{
			Vector R_vec = dx;
			R_vec.make_unit();//Unit of Resistance vector of particle because of attachement				
			Vector neg_R = R_vec*-1;
			//If the particle has any velocity, at this point the attachement brings it to a sudden halt.
			//Of course, this doesn't seem to arise very well from first principles.
			//It should be more dynamic. At every time step, the attachement should exert enough force to stop the particle.
			//As a temporary fix, this is ok.
			float vopp = vel.DotProduct(&neg_R);
			Vector vel_opp = neg_R * vopp;
			vel = vel - vel_opp;
		}
	}
}
void particle::applyForce(Vector force)
{
	if (vel.get_size() < max_speed)
	{
		

		float rem_force = 0;
		std::vector<int> active_attachements;

		for (int i = 0; i < attachements.size(); i++)
		{
			//Apply counter force for each attachement.

			
			Vector dx = attachements[i].pt.SubtractPointFromPoint(&part_loc); //points from particle to att. pt.
			float distance = dx.get_size();

			if (distance >= attachements[i].length)
			{
				//Since the particle is at the length of the attachment, the attachment might be active.
				//Check the force vector to see if it opposes attachement.
				Vector R_vec = dx;
				R_vec.make_unit();//Unit of Resistance vector of particle because of attachement				
				Vector neg_R = R_vec.MultiplyVectorByScalar(-1);
				float F_opp = force.DotProduct(&neg_R);							//Amount of opposing force along attachment axis

				if (F_opp > 0)
				{
					//Attachement is active.

					float diff = attachements[i].strength - F_opp;
					if (diff >= 0)
					{
						//Attachement holds. Opposing force is cancelled.
						Vector fres = R_vec.MultiplyVectorByScalar(F_opp);
						force += fres;
						//If the particle has any velocity, at this point the attachement brings it to a sudden halt.
						//Of course, this doesn't seem to arise very well from first principles.
						//It should be more dynamic. At every time step, the attachement should exert enough force to stop the particle.
						//As a temporary fix, this is ok.
						float vopp = vel.DotProduct(&neg_R);
						Vector vel_opp = neg_R * vopp;						
						vel = vel - vel_opp;
						rem_force = 0;
						break;
					}
					else
					{
						//Attachement is overcome.
						Vector fres = R_vec.MultiplyVectorByScalar(attachements[i].strength);
						force += fres;
						rem_force = -diff;
						active_attachements.push_back(i);
					}
				}

			}
			//Attachements break only if the force is able to overcome their combined strength.
			if (rem_force > 0)
			{
				//After all atachements have been evaluated, some force still remains. -> The active_attachements break.
				for (int i = 0; i < active_attachements.size(); i++)
				{
					int idx = active_attachements[i];
					if (idx > 0)
					{
						copy(attachements.begin() + idx, attachements.end(), attachements.begin() + idx - 1);
						attachements.resize(attachements.size() - 1);
					}
					else
					{
						attachements.clear();
					}
				}
			}

		}
		Vector da = force.DivideVectorByScalar(mass);
		acceleration = acceleration.AddVectorToVector(&da);
		if (int(acceleration.get_size() * 1000) == 0)
		{

			acceleration = acceleration * 0;
		}
	}
}

void particle::Collision(float mass_collide, float speed_x_collide, float speed_y_collide)
{
	if (!ignore_collision)
	{
		float final_speed_x = (vel.tuple[0] * mass + speed_x_collide * mass_collide - mass_collide * (vel.tuple[0] - speed_x_collide)) / (mass_collide + mass);
		float final_speed_y = (vel.tuple[1] * mass + speed_y_collide * mass_collide - mass_collide * (vel.tuple[1] - speed_y_collide)) / (mass_collide + mass);
		//final_speed_x = final_speed_x * -1;

		vel.tuple[0] = final_speed_x;
		vel.tuple[1] = final_speed_y;
	}

}

void particle::applyGravity(particle b)
{
	if (!gravity_exception)
	{
		float G = 200000.0;
		Vector disp = b.part_loc.SubtractPointFromPoint(&part_loc); //points from a to b.
		float distance = disp.get_size();
		disp = disp.DivideVectorByScalar(distance);
		float f = G * mass * b.mass / (distance * distance);
		float acc = f / mass;
		Vector a = disp.MultiplyVectorByScalar(acc);
		acceleration = acceleration.AddVectorToVector(&a);
	}

}

void particle::applyElectricity(particle b)
{
	if (!ignore_electrity)
	{
		Vector disp = b.part_loc.SubtractPointFromPoint(&part_loc);
		float distance = disp.get_size();
		disp = disp.DivideVectorByScalar(distance); //unit vector
		if (distance < (radius + b.radius))
		{
			float charge_product = charge * b.charge;
			if (charge_product > 0)
			{
				//then they have equipolar charge. Apply repellent force. The point of checking this is that I don't want to apply the force unless it's repellent. In the future, I'm sure I'll just use the regular equation to determine direction.

				if (gentle_electric_mode)
				{
					float E = 1.0f;
					float a_scalar = E * charge_product / (distance * distance) / mass;
					Vector a = disp.MultiplyVectorByScalar(a_scalar);

					vel = Vector(0, 0, 0);
					acceleration = acceleration.AddVectorToVector(&a);
				}
				else
				{
					float E = 100.0;
					float a_scalar = E * charge_product / (distance * distance) / mass;
					Vector a = disp.MultiplyVectorByScalar(a_scalar);
					acceleration = acceleration.AddVectorToVector(&a);
				}

			}
		}
	}

}




void particle::attach_to_point(myPoint pt, float str, float length)
{
	attachement nAtt;
	nAtt.pt = pt;
	nAtt.strength = str;
	nAtt.length = length;

	attachements.push_back(nAtt);
}



bool particle::inParticle(myPoint pt)
{
	Vector disp = pt.SubtractPointFromPoint(&part_loc);
	float dist = disp.get_size();
	if (dist <= radius)
	{
		return true;
	}
	else
	{
		return false;
	}
}

File: sim/particle.ts
import { Vector } from './vector'
import { Attachment } from './types'

let _nextId = 0

const G = 200000.0
const ELECTRIC_NORMAL = 100.0
const ELECTRIC_GENTLE = 1.0

export class Particle {
  id: string
  radius: number
  mass: number
  charge: number
  maxSpeed: number

  x: number; y: number; z: number
  vx: number; vy: number; vz: number
  ax: number; ay: number; az: number

  gravityException: boolean
  ignoreCollision: boolean
  ignoreElectricity: boolean
  gentleElectricMode: boolean

  attachments: Attachment[]

  constructor() {
    this.id = String(++_nextId)
    this.radius = 10
    this.mass = 1
    this.charge = 1
    this.maxSpeed = 100

    this.x = 0; this.y = 0; this.z = 0
    this.vx = 0; this.vy = 0; this.vz = 0
    this.ax = 0; this.ay = 0; this.az = 0

    this.gravityException = false
    this.ignoreCollision = false
    this.ignoreElectricity = false
    this.gentleElectricMode = false

    this.attachments = []
  }

  get speed(): number {
    return Math.sqrt(this.vx ** 2 + this.vy ** 2 + this.vz ** 2)
  }

  getVel(): Vector { return new Vector(this.vx, this.vy, this.vz) }
  setVel(v: Vector): void { this.vx = v.x; this.vy = v.y; this.vz = v.z }
  getAcc(): Vector { return new Vector(this.ax, this.ay, this.az) }
  setAcc(v: Vector): void { this.ax = v.x; this.ay = v.y; this.az = v.z }

  move(dt: number): void {
    this.x += this.vx * dt
    this.y += this.vy * dt
    this.z += this.vz * dt
  }

  accelerate(dt: number): void {
    this.vx += this.ax * dt
    this.vy += this.ay * dt
    this.vz += this.az * dt
    this.ax = 0; this.ay = 0; this.az = 0

    for (const att of this.attachments) {
      const dxv = new Vector(att.px - this.x, att.py - this.y, att.pz - this.z)
      const dist = dxv.magnitude()
      if (dist >= att.length) {
        const R = dxv.normalize()
        const negR = R.scale(-1)
        const vel = this.getVel()
        const vopp = vel.dot(negR)
        const velOpp = negR.scale(vopp)
        const newVel = vel.sub(velOpp)
        this.setVel(newVel)
      }
    }
  }

  applyForce(fx: number, fy: number, fz: number): void {
    if (this.speed >= this.maxSpeed) return

    let force = new Vector(fx, fy, fz)
    const active: number[] = []

    for (let i = 0; i < this.attachments.length; i++) {
      const att = this.attachments[i]
      const dxv = new Vector(att.px - this.x, att.py - this.y, att.pz - this.z)
      const dist = dxv.magnitude()

      if (dist >= att.length) {
        const R = dxv.normalize()
        const negR = R.scale(-1)
        const fOpp = force.dot(negR)

        if (fOpp > 0) {
          const diff = att.strength - fOpp
          if (diff >= 0) {
            // attachment holds — cancel opposing component and halt velocity along axis
            force = force.add(R.scale(fOpp))
            const vel = this.getVel()
            const vopp = vel.dot(negR)
            this.setVel(vel.sub(negR.scale(vopp)))
            break
          } else {
            // attachment is overcome
            force = force.add(R.scale(att.strength))
            active.push(i)
          }
        }
      }
    }

    // break overcome attachments
    if (active.length > 0) {
      this.attachments = this.attachments.filter((_, i) => !active.includes(i))
    }

    const da = force.divide(this.mass)
    if (Math.abs(da.magnitude() * 1000) > 0) {
      this.ax += da.x; this.ay += da.y; this.az += da.z
    }
  }

  applyForceVec(f: Vector): void {
    this.applyForce(f.x, f.y, f.z)
  }

  forceStop(): void { this.vx = 0; this.vy = 0; this.vz = 0 }

  collide(massOther: number, vxOther: number, vyOther: number): void {
    if (this.ignoreCollision) return
    const m1 = this.mass, m2 = massOther
    this.vx = (this.vx * m1 + vxOther * m2 - m2 * (this.vx - vxOther)) / (m1 + m2)
    this.vy = (this.vy * m1 + vyOther * m2 - m2 * (this.vy - vyOther)) / (m1 + m2)
  }

  applyGravity(b: Particle): void {
    if (this.gravityException) return
    const dx = b.x - this.x, dy = b.y - this.y, dz = b.z - this.z
    const dist = Math.sqrt(dx * dx + dy * dy + dz * dz)
    if (dist === 0) return
    const f = G * this.mass * b.mass / (dist * dist)
    const acc = f / this.mass
    this.ax += (dx / dist) * acc
    this.ay += (dy / dist) * acc
    this.az += (dz / dist) * acc
  }

  applyElectricity(b: Particle): void {
    if (this.ignoreElectricity) return
    const dx = b.x - this.x, dy = b.y - this.y, dz = b.z - this.z
    const dist = Math.sqrt(dx * dx + dy * dy + dz * dz)
    if (dist === 0) return
    if (dist < (this.radius + b.radius)) {
      const cp = this.charge * b.charge
      if (cp > 0) {
        const E = this.gentleElectricMode ? ELECTRIC_GENTLE : ELECTRIC_NORMAL
        const aScalar = E * cp / (dist * dist) / this.mass
        if (this.gentleElectricMode) {
          this.vx = 0; this.vy = 0; this.vz = 0
        }
        this.ax += (dx / dist) * aScalar
        this.ay += (dy / dist) * aScalar
        this.az += (dz / dist) * aScalar
      }
    }
  }

  attachToPoint(px: number, py: number, pz: number, strength: number, length: number): void {
    this.attachments.push({ px, py, pz, strength, length })
  }

  containsPoint(px: number, py: number, pz = 0): boolean {
    const dx = px - this.x, dy = py - this.y, dz = pz - this.z
    return Math.sqrt(dx * dx + dy * dy + dz * dz) <= this.radius
  }
}
